bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherEventListener.java

```diff
@@ -0,0 +1,352 @@
+        if (args.length > 0) {
+            if (args[0] instanceof String) {
+                reason = (String) args[0];
+            } else if (args[0] instanceof Exception) {
+                reason = String.format("Exception=%s Message=%s", args[0].getClass(),
+                        ((Exception) args[0]).getMessage());
+            }
+        }
+        logger.debug("Listener: Received socket event: {}, Reason: {}", event, reason);
+        bridgeHandler.markBridgeOffline(reason);
+    }
+
+    /*
+     * Parse the subscribed event, then tell the handler to update the station info channels.
+     */
+    private void handleSubscribed(String jsonData) {
+        logger.debug("Listener: subscribed={}", jsonData);
+        // Extract the station names and locations and give to handlers
+        try {
+            EventSubscribedJson subscribed = gson.fromJson(jsonData, EventSubscribedJson.class);
+            if (subscribed.invalidApiKeys != null) {
+                logger.debug("Listener: Invalid keys!! invalidApiKeys={}", subscribed.invalidApiKeys.toString());
+            }
+
+            if (subscribed.devices != null && subscribed.devices instanceof ArrayList) {
+                // Convert the ArrayList back to JSON, then parse it
+                String innerJson = gson.toJson(subscribed.devices);
+                DeviceJson[] stations = gson.fromJson(innerJson, DeviceJson[].class);
+
+                // Inform handlers of their name and location
+                for (DeviceJson station : stations) {
+                    logger.debug("Listener: Subscribed event has station: name = {}, location = {}, MAC = {}",
+                            station.info.name, station.info.location, station.macAddress);
+                    sendStationInfoToHandler(station.macAddress, station.info.name, station.info.location);
+                }
+            }
+            if (subscribed.isMethodSubscribe()) {
+                logger.debug("Listener: Subscribed to data events. Waiting for data...");
+            } else if (subscribed.isMethodUnsubscribe()) {
+                logger.debug("Listener: Unsubscribed from data events");
+            }
+        } catch (JsonSyntaxException e) {
+            logger.debug("Listener: Exception parsing subscribed.devices: {}", e.getMessage());
+            return;
+        }
+    }
+
+    /*
+     * Parse the weather data event, then send to handler to update the channels
+     */
+    private synchronized void handleData(String jsonData) {
+        logger.debug("Listener: Data: {}", jsonData);
+        try {
+            EventDataGenericJson data = gson.fromJson(jsonData, EventDataGenericJson.class);
+            if (StringUtils.isNotEmpty(data.macAddress)) {
+                sendWeatherDataToHandler(data.macAddress, jsonData);
+            }
+        } catch (JsonSyntaxException e) {
+            logger.info("Listener: Exception parsing subscribed event: {}", e.getMessage());
+            return;
```

Comment:
You have a lot of return's like this could you sweep some of them out :-) also for example on line 297

Reply:
Swept.